### DISCUSSION TITLE
Ask for feedback from professionally experienced C programmers

### DISCUSSION THREAD

**Discussion by stefanos82** [beejjorgensen/bgc#45] [Ideas]
Hey @beejjorgensen, I'm so glad you have enabled Discussions; things will get a lot better for your books.

As the title says, can we ask for professionally experienced C programmers for feedback on important topics we should know while learning C?

For example, 

- which ISO standard do they follow, if they do of course?
- what is the most frequent problem ( bug(s) vs feature(s) ) they face at work from colleagues that all C programmers should know?
- what is the most common error C authors do with their attempt to teach the language that confuses even the most experienced users? How would they teach it themselves that could lead to lots of "aha!" moments?

That's for a start.

**Comment by beejjorgensen** [comment]
I can comment a bit to get things going, but I'd love to hear from others on it.

Regarding standards, there was a time a while ago when folks would point out if you used a post-C89 feature since it might not build on an older compiler, but that doesn't seem to happen much any more for any standard.

My gut says that C99 is a common target nowadays, but the world has changed a lot since 2000 and I feel there's more keeping up-to-date with the newest compilers. Even if they're only coding for C99, the compiler could very likely be newer than that. I'm certain there are exceptions--a friend of mine works at a company where he's been trying to get them to upgrade for something ridiculous, like 10 years.

I have to confess, though, the last time I was paid to write C on a large codebase was 15 years ago (I started learning C in 1991 according to the price tag on my copy of K&R2.)

One thing I want to add to the reference section eventually is an indicator of what standard a function applies to so people know what they're getting into.

And I build all the samples with `gcc -Wall -Wextra -std=c18 -pedantic`, currently with GCC 10.2.0. (But that'll get upgraded as the new ones come out.)

**Comment by beejjorgensen** [comment]
And the one thing I saw from coworkers (and myself--though we were all reasonably adept) was accidentally stepping into undefined behavior. There's a lot of that in C.

In one case, I removed an unused local variable and it caused our completely-unrelated vector copy code to stop working in that function. A previous dev had been manually packing structs into a local char array. When I removed the local variable, it caused them to be misaligned on the stack and the vector copy required alignment. Changing the char array to an array of structs fixed it.

(Unfortunately the employee had since left, so I was unable to ask _why_ it was a char array instead of a more sensible struct array. But everything worked fine after the change.)

**Reply by stefanos82** [reply]
There are lots of things that got fixed in C11, let alone in C17, and we expect more things to get fixed as well in C2x when it gets finalized.

I do agree with you about C99 and the main reason is MSVC; they haven't fully supported C99 yet, therefore the book should respect that.

On the other hand, you could add any feature you want to teach by labeling it accordingly; for example, `_Static_assert` as C11 feature...you get the idea.

Also, you could teach about workarounds for those who cannot upgrade to C11 by using Jens Gustedt's [P99](https://gustedt.gitlabpages.inria.fr/p99/) which is a preprocessor for macros and functions for C99.

Basically it emulates C99 and C11; in some platforms it's even possible to emulate C17 features too!
